### PR TITLE
Cloud integration tests enhancements

### PR DESCRIPTION
**Pull request by danyilq**
* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- Improvement to the speed of cloud integration tests execution
- Makes cloud get_indexes request faster by not sending get_status call for each index initialisation and utlising instance mappings cache for it

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: tests/marqo_test.py [hunk 1]
```diff
@@ -228,7 +229,7 @@ def create_cloud_index(self, index_name, settings_dict=None, **kwargs):
         client = marqo.Client(**self.client_settings)
         index_name = f"{index_name}-{self.index_suffix}"
         if settings_dict or kwargs:
```
Comment (pandu-k) [inline tests/marqo_test.py]: can we do some validation? if settings_dict AND kwargs: throw error?

File: tests/marqo_test.py [hunk 2]
```diff
@@ -258,13 +260,16 @@ def create_test_index(self, index_name: str, settings_dict: dict = None, **kwarg
         return index_name
 
     def cleanup_documents_from_all_indices(self):
```
Comment (pandu-k) [inline tests/marqo_test.py]: Can you add a docstring highlighting this intended for use during cloud tests?

File: tests/marqo_test.py [hunk 3]
```diff
@@ -258,13 +260,16 @@ def create_test_index(self, index_name: str, settings_dict: dict = None, **kwarg
...
-                if client.http.get(f"/indexes/{index.index_name}/status")["index_status"] == "READY":
+            if index.index_name in self.indexes_to_cleanup:
+                print(f"Deleting documents from index {index.index_name}")
+                try:
+                    # veryfying that index is in the mapping
+                    self.client.config.instance_mapping.get_index_base_url(index.index_name)
                     docs_to_delete = [i['_id'] for i in index.search("", limit=100)['hits']]
                     while docs_to_delete:
                         index.delete_documents(docs_to_delete, auto_refresh=True)
                         docs_to_delete = [i['_id'] for i in index.search("", limit=100)['hits']]
```
Comment (pandu-k) [inline tests/marqo_test.py]: is it worth doing some error handling? so that we can detect if we're stuck in a potentially infinite loop (like the cloud has a bug preventing us from deleting docs?)

For example:
```
max_loops = 1000
if number_of_while_loops > max_loops:
    raise RuntimeError("Could delete all documents in index {index.index_name} after {max_loops} tries")
```

File: tests/marqo_test.py [hunk 4]
```diff
@@ -152,6 +153,8 @@ def setUpClass(cls) -> None:
         cls.generic_test_index_name = 'test-index'  # used as a prefix when index is created with settings
         cls.generic_test_index_name_2 = cls.generic_test_index_name + '-2'
 
+        cls.indexes_to_cleanup = set()
```
Comment (pandu-k) [inline tests/marqo_test.py]: Doesn't that mean there will be one of these dicts per test class? What about using a global dictionary?

File: tests/scripts/delete_all_cloud_test_indexes.py [hunk 5]
```diff
@@ -17,7 +17,7 @@ def delete_all_test_indices():
         if index.index_name.startswith('test-index'):
             if suffix is not None and suffix in index.index_name.split('-'):
                 if index.get_status()["index_status"] == marqo.enums.IndexStatus.CREATED:
-                    index.delete()
+                    index.delete(wait_for_readiness=False)
```
Comment (pandu-k) [inline tests/scripts/delete_all_cloud_test_indexes.py]: Should we do some waiting until we are confirmed deletion (including retrying + timeout if unsuccessful)? 

We should be careful about not-cleaned-up resources

File: tests/v0_tests/test_index.py
Comment (pandu-k) [inline tests/v0_tests/test_index.py:1]: more edge cases for health please

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    combined_str = json.dumps(dict_to_hash, sort_keys=True)
+    crc32_hash = zlib.crc32(combined_str.encode())
+    short_hash = hex(crc32_hash & 0xffffffff)[2:][
+                 :10]  # Take the first 10 characters of the hexadecimal representation
+    print(f"Created index with settings hash: {short_hash} for settings: {dict_to_hash}")
+    return short_hash
+
+
+def populate_indices():
+    test_uniqueness_id = os.environ.get("MQ_TEST_RUN_IDENTIFIER", "")
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py:48]: do we need to raise an error if we can't find the target env var?

File: tests/marqo_test.py [hunk 7]
```diff
@@ -122,12 +122,20 @@ def wrapper(self, *args, **kwargs):
...
+# replication of create_settings_hash from marqo.marqo_test
+def create_settings_hash(settings_dict, **kwargs):
     """
     Creates a hash from the settings dictionary and kwargs. Used to ensure that each index is created unique.
     Size is restricted on 10 characters to prevent having to big index name which could cause issues.
     """
-    dict_to_hash = settings_dict if settings_dict else kwargs
+    if settings_dict:
+        index_defaults = settings_dict.get("index_defaults", {}).copy()
+        if index_defaults and len(settings_dict) != 1:
```
Comment (pandu-k) [inline tests/marqo_test.py]: why check ` if ... settings_defaults != 1`?

Comment (pandu-k) [reply tests/marqo_test.py]: And why do we delete and re-add it?

File: tests/marqo_test.py [hunk 8]
```diff
@@ -122,12 +122,20 @@ def wrapper(self, *args, **kwargs):
...
+            del settings_dict["index_defaults"]
+            settings_dict.update(index_defaults)
+    print(settings_dict)
+    quit()
```
Comment (pandu-k) [inline tests/marqo_test.py]: i am guessing this is just for testing. ..

File: tests/marqo_test.py (hunk 2 above)
Comment (pandu-k) [reply tests/marqo_test.py]: And how it cleans up the indexes in `indexes_to_cleanup`

Also, I prefer if `indexes_to_cleanup` was given to this function as an argument

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        "cool-index" + '-' + test_uniqueness_id:
+            [{}],
+        "first-index" + '-' + test_uniqueness_id:
+            [{}],
+        "weight-index" + '-' + test_uniqueness_id:
+            [{}],
+        "mmodal-index" + '-' + test_uniqueness_id:
+            [{"treat_urls_and_pointers_as_images": True, "model": "ViT-B/32"}],
+        generic_test_index_name + '-' + test_uniqueness_id:
+            [{},
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py:54]: Can you add a docstring/comments about why a single index name maps to multiple index settings ?

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    api_key = os.environ.get("MARQO_API_KEY", None)
+    if api_key:
+        marqo_settings["api_key"] = api_key
+
+    mq = marqo.Client(**marqo_settings)
+
+    indexes_to_create = []
+
+    for index_name, configs_list in index_name_to_config_mappings.items():
+        for config in configs_list:
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py]: calling this config is confusing (I thought it was referring to an instance of the Config class). How about calling this index_configs, for example?

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+                     "ann_parameters": {
+                         "parameters": {
+                             "m": 24
+                         }
+                     }
+                 }
+             }
+              },
+             {"treat_urls_and_pointers_as_images": True,
+              "model": "ViT-B/32",
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py:66]: We definitely need to add to docstring/comments that if settings_dict is not included in the dicts, then it implies that the dict represents the index creation kwargs

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    mq = marqo.Client(**marqo_settings)
+
+    indexes_to_create = []
+
+    for index_name, configs_list in index_name_to_config_mappings.items():
+        for config in configs_list:
+            settings_dict = config.get("settings_dict")
+            if config:
+                if not settings_dict:
+                    index_name_to_create = index_name + '-' + create_settings_hash(settings_dict=None, **config)
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py]: is it possible to do `create_settings_hash(settings_dict=None, kwargs=**config)` for readability?

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+    return short_hash
+
+
+def populate_indices():
+    test_uniqueness_id = os.environ.get("MQ_TEST_RUN_IDENTIFIER", "")
+    generic_test_index_name = 'test-index'
+    index_name_to_config_mappings = {
+        generic_test_index_name + '-2' + '-' + test_uniqueness_id:
+            [{}],
+        "cool-index" + '-' + test_uniqueness_id:
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py]: why don't we have `create_cloud_index_name_base("cool-index", test_uniqueness_id)`?

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+    for index_name, configs_list in index_name_to_config_mappings.items():
+        for config in configs_list:
+            settings_dict = config.get("settings_dict")
+            if config:
+                if not settings_dict:
+                    index_name_to_create = index_name + '-' + create_settings_hash(settings_dict=None, **config)
+                else:
+                    index_name_to_create = index_name + '-' + create_settings_hash(settings_dict=settings_dict)
+            else:
+                index_name_to_create = index_name
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py:134]: Does this mean for those 4 demo indexes, it creates 4 different indexes? If their settings are the same they should be able to share the same cloud index, unless one test requires 2 different indexes to be set up at the same time,

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+        print(f"Creating {index_name} with config: {config}")
+        print(mq.create_index(index_name=index_name, wait_for_readiness=False, **config))
+
+    while True:
+        if all(creating_index in mq.config.instance_mapping._urls_mapping["READY"] .keys()
+               for creating_index, _ in indexes_to_create):
+            break
+        mq.config.instance_mapping._refresh_urls()
+        time.sleep(10)
+        print("Waiting for indexes to be created...")
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py:163]: Probably need to have some way to timeout and raise a runtime error

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+    dict_to_hash = settings_dict["index_defaults"] if settings_dict else kwargs
+    combined_str = json.dumps(dict_to_hash, sort_keys=True)
+    crc32_hash = zlib.crc32(combined_str.encode())
+    short_hash = hex(crc32_hash & 0xffffffff)[2:][
+                 :10]  # Take the first 10 characters of the hexadecimal representation
+    print(f"Created index with settings hash: {short_hash} for settings: {dict_to_hash}")
+    return short_hash
+
+
+def populate_indices():
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py:38]: shouldn't the mapping and, related functionality, of 
```
index_name <> settings <> cloud_testing_index_name 
```
Be available and easy to use for the entire test dir (rather than separating this logic into this file)?

File: tests/v0_tests/test_index.py
Comment (pandu-k) [inline tests/v0_tests/test_index.py:1]: more edge cases for the health endpoint

Comment (pandu-k) [inline tests/v0_tests/test_index.py:1]: Also, are we missing tests for any other new methods?

File: tests/marqo_test.py (hunk 4 above)
Comment (danyilq) [reply tests/marqo_test.py]: @pandu-k that's exactly what we want, purpose of this field is to reduce the amount of calls to delete_documents api. so with making it that way will make it so that class cleans up only those indices that were used during it's own tests.

File: tests/scripts/delete_all_cloud_test_indexes.py (hunk 5 above)
Comment (danyilq) [reply tests/scripts/delete_all_cloud_test_indexes.py]: @pandu-k if something goes wrong with deleting indexes it's already the same as tests failing, I think we should have warning for the integration tests account if some instance attached to it is running for more than 2 hours or it's up to us to decide.

File: tests/scripts/populate_indices_for_cloud_tests.py (hunk 14 above)
Comment (danyilq) [reply tests/scripts/populate_indices_for_cloud_tests.py:134]: @pandu-k yes, it could be correct, when I was applying changes I seen that we don't use generic name there and use 4 different names instead so I wasn't sure if it is important to keep them on completely fresh index and doc deletion might not help.

File: tests/marqo_test.py [hunk 17]
```diff
@@ -122,20 +122,18 @@ def wrapper(self, *args, **kwargs):
     return decorator
 
 
-# replication of create_settings_hash from marqo.marqo_test
 def create_settings_hash(settings_dict, **kwargs):
     """
     Creates a hash from the settings dictionary and kwargs. Used to ensure that each index is created unique.
     Size is restricted on 10 characters to prevent having to big index name which could cause issues.
     """
```
Comment (pandu-k) [inline tests/marqo_test.py]: ... we de-nest the index_defaults so that some kwarg indexes and setttings_dict indexes are more likely to generate the same index name, saving us from creating an unnecessary index:
```
settings_dict (original) = {"index_defaults": {"a": 1}, "number_of_shards": 2}
-> settings_dict (which we will use for hashing) = {"a": 1, "number_of_shards": 2}
# which matches the following index kwargs:
kwargs = {"a": 1, "number_of_shards": 2}
```

File: tests/marqo_test.py [hunk 18]
```diff
@@ -122,20 +122,18 @@ def wrapper(self, *args, **kwargs):
...
     if settings_dict:
+        settings_dict = settings_dict.copy()
```
Comment (pandu-k) [inline tests/marqo_test.py]: settings_dict_simplified = ... .copy()

File: tests/marqo_test.py [hunk 19]
```diff
@@ -266,17 +262,24 @@ def create_test_index(self, index_name: str, settings_dict: dict = None, **kwarg
...
+        idx = self.client.index(index_to_cleanup)
+        max_attempts = 1000
+        print(f"Deleting documents from index {idx.index_name}")
+        try:
+            # veryfying that index is in the mapping
+            idx.refresh()
+            self.client.config.instance_mapping.get_index_base_url(idx.index_name)
+            docs_to_delete = [i['_id'] for i in idx.search("", limit=100)['hits']]
+            attempt = 0
+            while docs_to_delete:
```
Comment (pandu-k) [inline tests/marqo_test.py]: Also consider: `index.get_stats()["documentCount"] > 0 ...`

Comment (pandu-k) [reply tests/marqo_test.py]: perhaps add refresh to loop

File: tests/scripts/delete_all_cloud_test_indexes.py (hunk 5 above)
Comment (pandu-k) [reply tests/scripts/delete_all_cloud_test_indexes.py]:
```
if (after waiting for an 20 min for all indexes to be deleted):
    raise RunTimeErrror(f"problem deleting indexes. please manually clean up indexes: {index-1, index-2}")

```

File: tests/scripts/populate_indices_for_cloud_tests.py [hunk 20]
```diff
@@ -6,12 +6,19 @@
 import marqo
 
 
+# replication of create_settings_hash from marqo.marqo_test
```
Comment (pandu-k) [inline tests/scripts/populate_indices_for_cloud_tests.py:9]: try to get the importing to work to remove duplicate code

File: tests/marqo_test.py [hunk 21]
```diff
@@ -235,11 +229,13 @@ def create_cloud_index(self, index_name, settings_dict=None, **kwargs):
         """
```
Comment (pandu-k) [inline tests/marqo_test.py:210]: We should throw an error if the cloud index doesn't yet exist

File: tests/marqo_test.py
Comment (pandu-k) [inline tests/marqo_test.py:1]: Add to docstring: "We will not actually create and delete real cloud indexes during this test suite, because this slows down py-marqo <> Marqo cloud tests. We should still test these methods with mocking. End-to-end creation and deletion tests will be done elsewhere."

File: tests/marqo_test.py (hunk 21 above)
Comment (pandu-k) [reply tests/marqo_test.py:210]: Make sure the  error message is helpful for new devs: they should know where to register their new index and index settings.

File: tests/scripts/run_cloud_tests.py [hunk 22]
```diff
@@ -1,19 +1,36 @@
...
+    signal.signal(signal.SIGINT, handle_interrupt)
 
-    import pytest
-    pytest_args = ['tests/', '-m', 'not ignore_during_cloud_tests'] + sys.argv[1:]
-    pytest.main(pytest_args)
-    print("All tests has been executed, proceeding to delete indices")
+    test_start_time = time.time()
+    try:
+        set_unique_run_identifier()
+        print(f"Using unique identifier: {os.environ.get('MQ_TEST_RUN_IDENTIFIER', '')}")
```
Comment (pandu-k) [inline tests/scripts/run_cloud_tests.py]: let's just fail if the env var is not set: `os.environ["MQ_TEST_RUN_IDENTIFIER"]`

File: tests/v0_tests/test_marqo_cloud_instance_mapping.py [hunk 23]
```diff
@@ -110,6 +110,17 @@ def test_request_of_creating_index_raises_error(self, mock_get):
         with self.assertRaises(MarqoCloudIndexNotReadyError):
             mapping.get_index_base_url("index2")
 
+    @patch("requests.get")
+    def test_modifying_state_returns_as_ready(self, mock_get):
```
Comment (pandu-k) [inline tests/v0_tests/test_marqo_cloud_instance_mapping.py:145]: do a test for logging (make sure we aren't getting warnings logged)

File: tests/v0_tests/test_index.py
Comment (pandu-k) [inline tests/v0_tests/test_index.py:1]: Add test for get_indexes() unable to update cache: we should log a warning but continue

File: tests/marqo_test.py [hunk 24]
```diff
@@ -241,11 +226,12 @@ def create_cloud_index(self, index_name, settings_dict=None, **kwargs):
...
             elif status != "READY":
-                self.client.create_index(index_name, settings_dict=settings_dict,
-                                         inference_node_type="marqo.CPU.large", storage_node_type="marqo.basic", **kwargs)
+                raise RuntimeError(f"Indexes must be precreated during cloud tests. Index {index_name} is not ready. "
+                                   f"Please specify it's settings in scripts/populate_all_indices.py")
         except (MarqoWebError, TypeError) as e:
-            self.client.create_index(index_name, settings_dict=settings_dict,
-                                     inference_node_type="marqo.CPU.large", storage_node_type="marqo.basic", **kwargs)
+            raise RuntimeError(f"Indexes must be precreated during cloud tests. Index {index_name} is not ready. "
+                               f"Please specify it's settings in scripts/populate_all_indices.py")
```
Comment (pandu-k) [inline tests/marqo_test.py]: nit, grammar: `its` not `it's`

File: setup.py [hunk 25]
```diff
@@ -18,7 +18,7 @@
         "tox"
     ],
     name="marqo",
-    version="1.2.2",
+    version="1.2.3",
```
Comment (pandu-k) [inline setup.py]: this needs to be bumped again. See here: https://pypi.org/project/marqo/

File: tests/v0_tests/test_marqo_cloud_instance_mapping.py [hunk 26]
```diff
@@ -134,6 +134,175 @@ def test_request_of_not_existing_index_raises_error(self, mock_get):
         with self.assertRaises(MarqoCloudIndexNotFoundError):
             mapping.get_index_base_url("index2")
 
+    @patch("requests.get")
+    def test_get_indexes_fails_cache_doesnt_update(self, mock_get):
```
Comment (pandu-k) [inline tests/v0_tests/test_marqo_cloud_instance_mapping.py:167]: this is a good test. What about a similar test, but the first time trying to `get_indexes` fails?

**Review by pandu-k — DISMISSED** [review]
